app/ldap/server/sync.js

```diff
@@ -175,14 +202,122 @@ export function getDataToSyncUserData(ldapUser, user) {
 		return userData;
 	}
 }
+export function mapLdapGroupsToUserRoles(ldapUser, user) {
+	const syncUserRoles = settings.get('LDAP_Sync_User_Data_Groups');
+	const syncUserRolesAutoRemove = settings.get('LDAP_Sync_User_Data_Groups_AutoRemove');
+	const syncUserRolesFieldMap = settings.get('LDAP_Sync_User_Data_GroupsMap').trim();
+
+	const roles = Roles.find({}, {
+		fields: {
+			_updatedAt: 0,
+		},
+	}).fetch();
+	const userRoles = [];
+	if (syncUserRoles && syncUserRolesFieldMap) {
+		const fieldMap = JSON.parse(syncUserRolesFieldMap);
+		_.map(fieldMap, function(userField, ldapField) {
+			const [roleName] = userField.split(/\.(.+)/);
+			if (!_.find(roles, (el) => el._id === roleName)) {
+				logger.debug(`User Role doesn't exist: ${ roleName }`);
+			} else {
+				logger.debug(`User role exists for mapping ${ roleName } -> ${ ldapField }`);
+
+				if (!isUserInLDAPGroup(ldapUser, user, ldapField)) {
+					if (syncUserRolesAutoRemove) {
+						const del = Roles.removeUserRoles(user._id, roleName);
+						if (settings.get('UI_DisplayRoles') && del) {
+							Notifications.notifyLogged('roles-change', {
+								type: 'removed',
+								_id: roleName,
+								u: {
+									_id: user._id,
+									username: user.username,
+								},
+							});
+						}
+					}
+				} else {
+					userRoles.push(roleName);
+				}
+			}
+		});
+	}
+	if (_.size(userRoles)) {
+		return userRoles;
+	}
+}
+export function createRoomForSync(channel) {
+	logger.info(`Channel '${ channel }' doesn't exist, creating it.`);
 
-
+	const room = createRoom('c', channel, settings.get('LDAP_Sync_User_Data_Groups_AutoChannels_Admin'), [], false, { customFields: { ldap: true } });
+	if (!room || !room.rid) {
+		logger.error(`Unable to auto-create channel '${ channel }' during ldap sync.`);
+		return;
+	}
+	room._id = room.rid;
+	return room;
+}
+export function mapLDAPGroupsToChannels(ldapUser, user) {
+	const syncUserRoles = settings.get('LDAP_Sync_User_Data_Groups');
+	const syncUserRolesAutoChannels = settings.get('LDAP_Sync_User_Data_Groups_AutoChannels');
+	const syncUserRolesEnforceAutoChannels = settings.get('LDAP_Sync_User_Data_Groups_Enforce_AutoChannels');
+	const syncUserRolesChannelFieldMap = settings.get('LDAP_Sync_User_Data_Groups_AutoChannelsMap').trim();
+
+	const userChannels = [];
+	if (syncUserRoles && syncUserRolesAutoChannels && syncUserRolesChannelFieldMap) {
+		const fieldMap = JSON.parse(syncUserRolesChannelFieldMap);
+
+		_.map(fieldMap, function(channels, ldapField) {
+			if (typeof channels === 'object') {
```

Comment:
Should be `if (Array.isArray(channels))` and could do `channels = [channels]` if it's not an array to prevent code duplication below